crates/ide-completion/src/context.rs

```diff
@@ -155,13 +155,54 @@ pub(crate) struct ExprCtx {
 pub(crate) enum TypeLocation {
     TupleField,
     TypeAscription(TypeAscriptionTarget),
-    GenericArgList(Option<ast::GenericArgList>),
+    /// Generic argument position e.g. `Foo<$0>`
+    GenericArg(Option<(ast::GenericArgList, Option<hir::Trait>, Option<ast::GenericParam>)>),
+    /// Associated type equality constraint e.g. `Foo<Bar = $0>`
+    AssocTypeEq,
+    /// Associated constant equality constraint e.g. `Foo<X = $0>`
+    AssocConstEq,
     TypeBound,
     ImplTarget,
     ImplTrait,
     Other,
 }
 
+impl TypeLocation {
+    pub(crate) fn complete_lifetimes(&self) -> bool {
+        match self {
+            TypeLocation::GenericArg(Some((_, _, Some(param)))) => {
+                matches!(param, ast::GenericParam::LifetimeParam(_))
+            }
+            _ => false,
+        }
+    }
+
+    pub(crate) fn complete_consts(&self) -> bool {
+        match self {
+            TypeLocation::GenericArg(Some((_, _, Some(param)))) => {
+                matches!(param, ast::GenericParam::ConstParam(_))
+            }
+            TypeLocation::AssocConstEq => true,
+            _ => false,
+        }
+    }
+
+    pub(crate) fn complete_types(&self) -> bool {
+        match self {
+            TypeLocation::GenericArg(Some((_, _, Some(param)))) => {
+                matches!(param, ast::GenericParam::TypeParam(_))
+            }
```

Comment:
```suggestion
            TypeLocation::GenericArg(Some((_, _, Some(ast::GenericParam::TypeParam(_))))) => true,
```

Reply:
This would make the match return `true` for `TypeLocation::GenericArg(Some((_, _, Some(ast::GenericParam::ConstParam(_)))))` because of the wildcard arm below, which we don't want